torch/autograd/_functions/reduce.py

```diff
@@ -71,7 +71,10 @@ def safe_zeros_backward(inp, dim):
             exclusive_normal = exclusive_normal_nocp.cumprod(dim)
 
             def reverse_dim(var, dim):
-                return var.index_select(dim, Variable(torch.arange(var.size(dim) - 1, -1, -1)).long())
+                index = Variable(torch.arange(var.size(dim) - 1, -1, -1)).long()
+                if var.is_cuda:
+                    index = index.cuda()
```

Comment:
Yes, it does create a tensor on the same device as `var`. But the `arange` operation fails when the output tensor is not on the current cuda device.
Anyway, the CUDA device id should not be a concern at such a low level in my opinion, these basic operations are performed on the same GPU card.